DigitalLearningSolutions.Web.Tests/Controllers/Register/ClaimAccountControllerTests.cs

```diff
@@ -171,63 +188,216 @@ public void CompleteRegistrationGet_with_logged_in_user_redirects_to_LinkDlsAcco
-                .WithRouteValue("centreName", DefaultCentreName)
-                .WithRouteValue("candidateNumber", DefaultCandidateNumber);
+                .WithRouteValue("email", model.Email)
+                .WithRouteValue("centreName", model.CentreName)
+                .WithRouteValue("candidateNumber", model.CandidateNumber);
+        }
+
+        [Test]
+        public async Task CompleteRegistrationWithoutPassword_with_email_in_use_returns_NotFound()
+        {
+            // Given
+            var model = GivenClaimAccountViewModel();
+
+            A.CallTo(() => userDataService.PrimaryEmailIsInUse(model.Email)).Returns(true);
+
+            // When
+            var result = await controller.CompleteRegistrationWithoutPassword(
+                model.Email,
+                model.RegistrationConfirmationHash
+            );
+
+            // Then
+            ACallToConvertTemporaryUserToConfirmedUserMustNotHaveHappened();
+
+            result.Should().BeNotFoundResult();
         }
 
         [Test]
-        public void CompleteRegistrationPost_with_email_in_use_returns_NotFound()
+        public async Task CompleteRegistrationWithoutPassword_with_password_not_set_by_admin_returns_NotFound()
         {
             // Given
-            GivenClaimAccountViewModel();
+            var model = GivenClaimAccountViewModel(wasPasswordSetByAdmin: false);
+
+            // When
+            var result = await controller.CompleteRegistrationWithoutPassword(
+                model.Email,
+                model.RegistrationConfirmationHash
+            );
 
-            A.CallTo(() => userDataService.PrimaryEmailIsInUse(DefaultEmail)).Returns(true);
+            // Then
+            result.Should().BeNotFoundResult();
+        }
+
+        [Test]
+        [TestCase(null, null)]
+        [TestCase("   ", null)]
+        [TestCase(null, "   ")]
+        [TestCase("   ", "   ")]
+        [TestCase(null, DefaultCode)]
+        [TestCase("   ", DefaultCode)]
+        [TestCase(DefaultEmail, null)]
+        [TestCase(DefaultEmail, "   ")]
+        public async Task CompleteRegistrationWithoutPassword_with_invalid_email_or_code_redirects_to_AccessDenied(
+            string email,
+            string code
+        )
+        {
+            // Given
+            var model = GivenClaimAccountViewModel(email: email, registrationConfirmationHash: code);
 
             // When
-            var result = controller.CompleteRegistrationPost(DefaultEmail, DefaultCode);
+            var result = await controller.CompleteRegistrationWithoutPassword(
+                model.Email,
+                model.RegistrationConfirmationHash
+            );
 
             // Then
+            result.Should().BeRedirectToActionResult().WithControllerName("LearningSolutions")
+                .WithActionName("AccessDenied");
+        }
+
+        [Test]
+        public async Task CompleteRegistrationWithoutPassword_with_no_existing_user_redirects_to_AccessDenied()

```

Comment:
Is this for the case where the email and hash don't match any user-to-be-claimed?

Reply:
Yes. I've renamed these tests `*_with_no_user_to_be_claimed_redirects_to_AccessDenied` and added another private method: `GivenEmailAndCodeDoNotMatchAUserToBeClaimed`

Reply:
Very nice!